stdlib/public/core/NativeDictionary.swift

```diff
@@ -704,6 +704,30 @@ extension _NativeDictionary { // Deletion
     _storage._count = 0
     invalidateIndices()
   }
+  
+  @inlinable
+  mutating func remove(
+    where predicate: (Key) throws -> Bool,
+    keepCapacity: Bool = false
+  ) rethrows {
+    //FIXME(performance): it would be more efficient to filter as we go, instead
+    //of copying (if not unique) and then filtering
+    ensureUnique(isUnique: isUnique, capacity: capacity)
+    var removedCount = 0
+    for bucket in hashTable where hashTable.isOccupied(bucket) {
+      let key = self.uncheckedKey(at: bucket)
+      if try predicate(key) {
+        hashTable.delete(at: bucket, with: self)
```

Comment:
So my understanding is that linear addressing makes this sequential deletion tricky… Let's say `a`, `b`, `c`, and `d` are added in order, and all hash to the same value (mod 4):

```
┌───┬───┬───┬───┐
│ a │ b │ c │ d │
└───┴───┴───┴───┘
  0   1   2   3 
```

If `predicate` returns `true` for all of these, this will delete `a` at position 0, which then requires shifting the other three elements down. Then it will move on to position 1 without considering `b`, and delete `c`, which will shift `d` down, leaving this at the end:

```
┌───┬───┬───┬───┐
│ b │ d │   │   │
└───┴───┴───┴───┘
  0   1   2   3 
```

Reply:
Ah right. We can’t actually avoid rehashing the way I naively hoped. Alas

Reply:
Actually wait, no. Isn’t this what deletion tombstones are intended to prevent? Do we not do that in our hash table?

Reply:
No, our hash table eagerly shuffles elements back until it finds a hole instead of using tombstones. I think this will be correct if you step from bucket to bucket manually, so you can recheck the current bucket if it's still occupied after deletion — you can use `startBucket`, `endBucket`, and `occupiedBucket(after:)` for that. (That will be faster anyway, since you'll be able to skip empty regions a word at a time instead of a bit at a time.)